First day with LibraryViewManager

jmigual opened the pull request:
By the way now that I've seen your last message in the Mixxx devel, forget about the `getButtonBar()`, `getLeftPane()` and `getRightPane()` it must be `set*()` and the LegacySkinParser must setup and style them.

daschuer commented on src/library/libraryviewmanager.cpp:
```diff
@@ -0,0 +1,60 @@
+    m_pLeftPane = new QStackedWidget;
+    m_features.resize(RIGHT_PANE_COUNT);
+
+    for (int i = 0; i < RIGHT_PANE_COUNT; ++i) {
```
Yes, this should be done by the Skin, similar how it is working today.

daschuer commented:
IMHO you are on th right track, except the skin interaction.

jmigual commented:
Hi Daniel,
I've made some changes if you want to review them

daschuer commented on src/library/library.cpp:
```diff
@@ -137,13 +92,11 @@ Library::Library(QObject* parent, UserSettingsPointer pConfig,
-        delete feature;
+    
+    for (LibraryFeature* f : m_features) {
+        delete f;
```
OK this looks scary, but works, because of the clear below. 
How about using qDeleteAll or use smart pointers.

jmigual replied:
Ok, I'll use qDeleteAll here a smart pointer is not necessary, he have full knowledge of the pointer behaviour

daschuer commented on src/library/library.cpp:
```diff
@@ -377,3 +345,69 @@ void Library::slotSetTrackTableRowHeight(int rowHeight) {
 }
+
+void Library::libraryWidgetFocused() {
+    WLibrary* pane = dynamic_cast<WLibrary*>(sender());
```
use qobject_cast<>
Here some hints why: 
https://forum.qt.io/topic/59498/qobject_cast-vs-dynamic_cast/4

jmigual replied:
Ok, fine

daschuer commented on src/widget/wbuttonbar.cpp:
```diff
@@ -0,0 +1,25 @@
+void WButtonBar::addButton(const QIcon& icon, const QVariant& title, const QString& data) {
+    QPushButton* button = new QPushButton(icon, title.toString(), this);
+
+    m_pButtonGroup->addButton(button, m_data.size());
```
just m_pButtonGroup->addButton(button) appends to the end.

daschuer replied:
Ignore my comment, the code is just right.

daschuer commented:
Good start. 
From our new model, does the library still need to track the focus?

jmigual commented:
Yes it still needs to track the focus because the only pane that must change is the focused pane. By the way for me it seems that `WLibrary` does the things that the `LibraryPaneManager` should do so I used this class and altered a bit to fit the new concept.

daschuer commented on src/library/library.cpp:
```diff
@@ -172,24 +125,42 @@ void Library::bindSidebarWidget(WLibrarySidebar* pSidebarWidget) {
+    }
+}
+
+void Library::bindLibraryWidget(WLibrary* pLibraryWidget,
```
I think we need to consolidate the names again. 
The term "Pane" works for me, but i think it conflicts with the "Library" which is a fuzzy word in our context. 

He have currently: 
* LibrarySideBar 
* LibrarySideBarExpandet 
* Library (=Pane) 

How about use LibraryPane?  Than all words have the Library prefix, which can be used or not, depending of the context. 

This mean in this context, we will have Library::bindPaneWidget()

jmigual replied:
Ok it seems good to me

daschuer commented on src/library/library.cpp:
```diff
@@ -199,30 +170,46 @@ void Library::bindWidget(WLibrary* pLibraryWidget,
+    */
+    
+    if (! m_panes.contains(id)) {
+        createPane(id);
```
This construct can be optimized to something like 
LibraryPaneManager* pane = getOrCreatePane(id);
to avoid double hash lookups.

daschuer commented on src/library/library.cpp:
```diff
@@ -377,3 +378,87 @@ void Library::slotSetTrackTableRowHeight(int rowHeight) {
+
+LibraryPaneManager *Library::getFocusedPane() {
+    qDebug() << "Focused" << m_focusedPane;
+    if (m_focusedPane == -1) {
```
Do we have a grantee that the skin designed does not use -1 as ID?

jmigual replied:
Yes, in the LegacySkinParser I if the skin designer uses a number below 0 it will set the id to 0 and will show a warning

jmigual replied:
Now I've noticed that I haven't coded this and I only thought about it... I'm a genius... :sweat:

daschuer commented on src/library/library.cpp:
```diff
@@ -377,3 +378,87 @@ void Library::slotSetTrackTableRowHeight(int rowHeight) {
+LibraryPaneManager *Library::getFocusedPane() {
+    qDebug() << "Focused" << m_focusedPane;
+    if (m_focusedPane == -1) {
+        return m_pSidebarExpanded;
```
This looks like it breaks our concept. Is m_pSidebarExpanded also a Pane? 
Will all users of getFocusedPane() expect that the return value is m_pSidebarExpanded?

jmigual replied:
Yes it's also a pane because it have the same behavior as the other panes, I noticed that I was adding the same functions twice... There will be some methods that must be aware of the focused pane being the m_pSidebarExpanded and others that won't need this option... But with this comment I've noticed some bugs I have to fix, thanks!

daschuer replied:
Not sure if I am correct, but we will probably have some common functions and some individuals for LibraryPane and SidbarExpanded, so this could be a candidate for a subclass with two childs. 
.. we will see later, I do not want to stop your good pace ;-)

daschuer commented on src/library/librarypanemanager.h:
```diff
@@ -0,0 +1,69 @@
+
+  public:
+
+    const int RIGHT_PANE_COUNT = 2;
```
This seams to be unused now. By the way: we would use kRightPaneCount as name.

daschuer commented on src/library/librarypanemanager.h:
```diff
@@ -0,0 +1,69 @@
+    
+    void focused();
+    
+    void showTrackModel(QAbstractItemModel* model);
```
The feature has not necessarily Tracks.

jmigual replied:
You're right, but if the Pane has tracks it must have the bindTrackTable to connect the `showTrackModel()`, `searchStarting()` and `onSearchCleared()` signals to the track table. Maybe this can be done in the `Library` instead of the `LibraryPaneManager`... :confused:

daschuer replied:
Or in the feature .. only it nows if it has a track table or not.
But we can discuss it later.

daschuer commented on src/library/libraryviewfeature.h:
```diff
@@ -0,0 +1,41 @@
+    virtual QString getDefaultNameView() = 0; 
+
+    // A LibraryViewFeature always has a right and a left pane
+    virtual QWidget* getLeftPane() = 0;
```
getSidebarExpanded()

jmigual replied:
Oh this is an old class, it should be removed...

daschuer commented on src/library/libraryviewfeature.h:
```diff
@@ -0,0 +1,41 @@
+    // A LibraryViewFeature always has a right and a left pane
+    virtual QWidget* getLeftPane() = 0;
+
+    virtual QWidget* getRightPane() = 0;
```
getPaneWidget()

I just notice there is an other word race. 
Pane can be a PaneWiget or the PaneManager ... 
Should we add a suffix of a prefix to all feature widgets?

daschuer commented on src/skin/legacyskinparser.cpp:
```diff
@@ -1135,21 +1138,25 @@ QWidget* LegacySkinParser::parseSpinny(const QDomElement& node) {
+        
+        if (id < 0) {
+            SKIN_WARNING(node, *m_pContext) << "Id must be >= 0";
+            id = 0;
```
IMHO it is better to skip the node if there is a wrong ID. This will make the skin error more obvious.

jmigual replied:
Ok

daschuer commented on src/library/librarypanemanager.cpp:
```diff
@@ -0,0 +1,99 @@
+    }
+    
+    // Since this event filter is for the entire application (to handle the
+    // mouse event), NEVER return true
```
because that ... ¿?

jmigual replied:
There are some cases where Qt does not consider a focus change but we do. The main example is with the main Crates/Playlists view which has a QTextBrowser that does not handle FocusIn / FocusOut events. Currently this only handles event change with mouse but I'll add later, keyboard and touch events too.

daschuer replied:
Ok, Thank you. 
Could you extend the commend here and explain what would happen if we return true? this would make the comment complete.

daschuer commented on src/library/library.h:
```diff
@@ -134,6 +140,15 @@ class Library : public QObject {
+    LibraryPaneManager* m_pSidebarExpanded;
+    QList<LibraryFeature*> m_features;
+    
+    // -1 for the Sidebar Expanded and >= 0 for the other widgets
```
a const definition for this would be nice

jmigual replied:
Now this is changed, it won't be necessary as there's a `bool` for the sidebar expanded and a `int` for the panes.

daschuer commented on src/library/cratefeature.cpp:
```diff
@@ -25,6 +25,8 @@
 #include "util/dnd.h"
 #include "util/time.h"
 
+const QString CrateFeature::m_sCrateViewName = QString("CRATEHOME");
```
We use k for const expressions and p for pointers. We have no rule for other type specific prefixes, this is IMHO not necessary with todays IDES. 

You may also put such things into an anonymous namespace to hide it from the linker.

jmigual replied:
Then I remove de 's' prefix and add it to an anonymous namespace?

jmigual commented:
Ok, it's starting to work. I've added the changes into the Deere skin so If you want try it and tell me your opinion (Sidebar Expanded select it's still not working).

daschuer commented:
Good news. I should find time tomorrow.

daschuer commented:
I have just tested this branch and I am impressed. 
Is there anything special to test?

There seams to be an focus issue with the SideBarExpanded. IMHO it should always change when selecting a new feature by button. The pane focus should only take the LibraryTrack tables into account. 

IMHO there should be no scroll bar on the LibrarySideBar by default. It is better to have the favorit buttons always on the same position. That could be issued by a flyout menu or a second temporary bar or a "more" button like on the google services menu on the default google search page. 

Do you have a good idea to make the buttons smaller? Something like a smaller font or / and a fixed wide + text ellipsis + tooltip for the full text. or only tooltips or two rows of text ..

jmigual commented:
> IMHO it should always change when selecting a new feature by button. 

Ok, I'll change it currently only changes if there's something to show but you are right it's a bit annoying.

> IMHO there should be no scroll bar on the LibrarySideBar by default. It is better to have the favorit buttons always on the same position. That could be issued by a flyout menu or a second temporary bar or a "more" button like on the google services menu on the default google search page.

I'll add a "more" button like the google services one but this will be done after everything else is working, currently there are still a few bugs.

> Do you have a good idea to make the buttons smaller?

Maybe a smaller font (can be configured with CSS), also putting the text on the side of the button instead of below also reduces considerably the size.

daschuer commented:
Ok, nice, the focus changing and the feature switching works like discussed. 

However, for me there is still a usability issue. I would like to be able to switch fast between the AutoDJ controls and other features, without ending up with too AutoDJ track table or none. 
What could be a solution? An extra show track table button, or a smart logic that prefers to keep a library pane layout?  

Other pending issues: 

* The AutoDJ buttons are squeezed when the pane is lowered by the preview player or the cover art. 
 I think we need a minimum size + scrollbars. 
* The AutoDJ crates are gone. 
* There is no library root node
* right click on the empty area in the playlist/crates LibrarySidebarExpanded should show the context menu.
* Browse / Rhythmbox ... missing treeview. 
* How about the explaining root view of the playlist and crate features? Any idea? remove?

jmigual commented:
> However, for me there is still a usability issue. I would like to be able to switch fast between the AutoDJ controls and other features, without ending up with too AutoDJ track table or none. 
What could be a solution? An extra show track table button, or a smart logic that prefers to keep a library pane layout?

I think that I don't understand this use case... 

> - There is no library root node

Is it necessary? clicking in the Library button in the Sidebar does the same as the Library root node

> - Right click on the empty area in the playlist/crates LibrarySidebarExpanded should show the context menu.

Totally agree

> - How about the explaining root view of the playlist and crate features? Any idea? remove?

What do you mean by this?

daschuer commented:
>> However, for me there is still a usability issue. I would like to be able to switch fast between the AutoDJ controls and other features, without ending up with too AutoDJ track table or none. 
What could be a solution? An extra show track table button, or a smart logic that prefers to keep a library pane layout?

> I think that I don't understand this use case... 

The use case is Auto DJ driven DJ-ing. Auto DJ is always on and the DJ sorts new tracks to the AutoDJ queue. In case he wants to adjust the fade-time, he can't just click on the AutoDJ button, because he may has the Library track table in focus, if he clicks anyway, he ends up with two times AutoDJ queue (most likely unwanted). If he has managed to select the right focus pane before, and wants continue browsing for tracks, he has to switch focus again to not hide the AutoDJ queue from one pane. 

It would be great if we can find a solution where the user can click around and the focus is switched automatically. Maybe we have to rethink our current focus idea, to match common tasks a bit better. Any ideas?   
   

>> - There is no library root node

> Is it necessary? clicking in the Library button in the Sidebar does the same as the Library root node

Not in any case. But IMHO we need an obvious solution to show the whole library again from the SideBarExpanded, because the user has switched to the missing Tracks here, so he should be able to switch back in a similar way.  

>> - How about the explaining root view of the playlist and crate features? Any idea? remove?

> What do you mean by this?

If you click on the crates node in the Mixxx release, the track table shows an explaining text and a Photo of a Crate. What will happen with this view in your design?

daschuer commented:
The minimum track table width is too big. IMHO we should allow to move the splitter too 0 width, or a reasonable small value near 0.

daschuer commented:
The button bar should have a max width.

daschuer commented:
All context menu items are grayed out.

daschuer commented:
When I repeat pressing the Library button, the selection in the Track table seams to jump around.
This happens also if I focus the other pane.

jmigual commented:
>>> However, for me there is still a usability issue. I would like to be able to switch fast between the AutoDJ controls and other features, without ending up with too AutoDJ track table or none. 
What could be a solution? An extra show track table button, or a smart logic that prefers to keep a library pane layout?

>> I think that I don't understand this use case...

> The use case is Auto DJ driven DJ-ing. Auto DJ is always on and the DJ sorts new tracks to the AutoDJ queue. In case he wants to adjust the fade-time, he can't just click on the AutoDJ button, because he may has the Library track table in focus, if he clicks anyway, he ends up with two times AutoDJ queue (most likely unwanted). If he has managed to select the right focus pane before, and wants continue browsing for tracks, he has to switch focus again to not hide the AutoDJ queue from one pane.

> It would be great if we can find a solution where the user can click around and the focus is switched automatically. Maybe we have to rethink our current focus idea, to match common tasks a bit better. Any ideas?

Ok now I understand it, what about adding something like CRTL+click in the button bar, with this we will now that only the sidebar expanded must change.

>>>There is no library root node

>>Is it necessary? clicking in the Library button in the Sidebar does the same as the Library root node

>Not in any case. But IMHO we need an obvious solution to show the whole library again from the SideBarExpanded, because the user has switched to the missing Tracks here, so he should be able to switch back in a similar way.

Ok I'll add a TreeItem in the Library tree to make more evident to the user how to return to the library.

>>>How about the explaining root view of the playlist and crate features? Any idea? remove?

>>What do you mean by this?

>If you click on the crates node in the Mixxx release, the track table shows an explaining text and a Photo of a Crate. What will happen with this view in your design?

This will remain there's no need to remove it.

> The minimum track table width is too big. IMHO we should allow to move the splitter too 0 width, or a reasonable small value near 0.

I'm on it as there are still some widgets that are not tables there they block the resize option...

> The button bar should have a max width.

Done

> All context menu items are grayed out.

In which case?

> When I repeat pressing the Library button, the selection in the Track table seams to jump around.
This happens also if I focus the other pane.

I cannot reproduce :cry:

daschuer commented:
> Ok now I understand it, what about adding something like CRTL+click in the button bar? With this we will now that only the sidebar expanded must change.

Nice Idea. But I would do it the other way round: If you click on the AutoDJ button and there is already a AutoDJ track table on the not focused pane, do not display a new one unless the user is pressing Ctrl. 

For the other features, we have already discussed that the LibraryPane remains unchanged when clicking on the button (the drop target discussion).  This can be sufficient for a feature change here. 
If a feature was active before and already occupies a pane, we can should reuse it unless the user is holding Ctrl. 

What do you think about it?

daschuer commented:
>>    When I repeat pressing the Library button, the selection in the Track table seams to jump around.
>>    This happens also if I focus the other pane.

> I cannot reproduce :cry: 

Ah now I got it, the randomize sorting kicks in. This can be avoided by not refreshing the library pane when the button is clicked again.

daschuer commented:
> Now that the splitter is gone, the Library buttons are chopped in the German version.

How about reintroduce the splitter along with a size aware stack. Small width = button bar without texts big size button bar with text?

daschuer commented:
I am stating to like the new layout. Good work.

jmigual commented:
>>Ok now I understand it, what about adding something like CRTL+click in the button bar? With this we will now that only the sidebar expanded must change.

>Nice Idea. But I would do it the other way round: If you click on the AutoDJ button and there is already a AutoDJ track table on the not focused pane, do not display a new one unless the user is pressing Ctrl.

>For the other features, we have already discussed that the LibraryPane remains unchanged when clicking on the button (the drop target discussion). This can be sufficient for a feature change here. 
If a feature was active before and already occupies a pane, we can should reuse it unless the user is holding Ctrl.

>What do you think about it?

Ok then if the user clicks the AutoDJ button and there's an AutoDJ table showing the library must only show the AutoDJ sidebar, is ok? 
And now I do not understand the other features, if a feature is showing and the user clicks the button sidebar then we do not do anything? If the feature is not shown then we display both table and feature's sidebar?

jmigual commented:
>>Now that the splitter is gone, the Library buttons are chopped in the German version.

>How about reintroduce the splitter along with a size aware stack. Small width = button bar without texts big size button bar with text?

I like the idea

jmigual commented:
>>>When I repeat pressing the Library button, the selection in the Track table seams to jump around.
This happens also if I focus the other pane.

>>I cannot reproduce :cry:

>Ah now I got it, the randomize sorting kicks in. This can be avoided by not refreshing the library pane when the button is clicked again.

Ok, I'll solve it

daschuer commented:
I am not sure If we have the same idea of the issue. Since the issue is getting tricky, it would be nice to know how you feel about this and if you could reproduce the issues on your side. 
I have noticed it by trying to simulate AutoDJ usage, in a typical stressed situation where AutoDJ is running out of tracks and a topmost queue track does not fit and will likely empty the dance floor. 
The DJ browse the library, noticed the bad track is loaded to a deck. 
Now he has to press "skip next", in time and find a suitable successor track. 

> Ok then if the user clicks the AutoDJ button and there's an AutoDJ table showing the library must only show the AutoDJ sidebar, is ok? 

Yes. 

> And now I do not understand the other features, if a feature is showing and the user clicks the button sidebar then we do not do anything? 

Yes, this will avoid the double randomize issue. In case the LibrarySideBarExpanded shows a different feature, only the sidebar is changed. 

> If the feature is not shown then we display both table and feature's sidebar?

I am unsure, but here is an idea to discuss: 

It depends on the feature (so it should be finally handled in the feature code). If the feature has only one LibraryPane view, like the AutoDJ it is shown on sidbar button click. If it has many views (like the playlist feature), switch able from the expanded sidebar, the LibrayPanes are not changed. 
The focused LibraryPane is changed once the user clicks inside the extended sidebar (e.g. on a specific play list). 

If a Library pane is already associated to the clicked feature, this is adopted as the focus widget. 

Implementation Idea: 

How about track a "featureFocus" and a "currentFocus" like that: 

* Start Mixxx
* Libraryfetaure is enabled with  "Libraryfetaure featureFocus" and  "currentFocus" at Pane 1
* Click on Pane 2: "Libraryfetaure featureFocus" = 1 and  "currentFocus" = 2 
* Enable AutoDJ feature. This adopts the currentFocus as "AutoDJ featureFocus" = 2
* Switch back to  LibraryFetaure, currentFocus = 1 adopting stored "Libraryfetaure featureFocus" = 1
* Switch back to AutoDJ, currentFocus = 2 adopting stored "AutoDJ featureFocus" = 1

How do you think about it?

jmigual commented:
Ok, I think I understand it, we can use this "feature focus" idea and also add a signal `switchToNotFocusedView` that allows the features to only show a view if it's not already shown in some pane. Tthen the existing `switchToView` signal will always show the table in the current focused pane unless it's already shown.

jmigual commented:
Ok, now it seems solved, I also added the new look into LateNight skin and added an evident focus change selector, now the selected pane has a border around it

daschuer commented:
Yeehaw :-) it feels good now. 

The focus indicator helps a lot. Together with the upcoming breadcrumb, it will be great. 

* Not all features like the History view play the new focus game now. 
* We also need to remove the default pane with the explaining text of some features to be able to use it as drop target. Any idea where to move it? 

Idea: 
* with a splitter into the expandable pane
* as default expandable pane when the tree is empty 
* A root node "Help"
* remove

jmigual commented:
Why do we have to remove it, can it be converted to a drop target without removing the explaining text?

daschuer commented:
In the single pane case, we drag from the track table and hover to the button bar. 
Now the explaining pane replaces the track table, which is not desired. 
I think we have discussed only show the LibrarySidebarExpanded in this case.  
.. or do you plan to distinguish the hover case from the click case?

daschuer commented on src/library/librarypanemanager.cpp:
```diff
@@ -0,0 +1,123 @@
+    connect(this, SIGNAL(switchToView(const QString&)),
+        m_pPaneWidget, SLOT(switchToView(const QString&)));
+
+    switch (pane) {
```
We should subclass LibraryPaneManager, to get rid of this switch.

jmigual replied:
Ok, I'll change it

daschuer commented on src/library/librarypanemanager.cpp:
```diff
@@ -0,0 +1,123 @@
+            break;
+        case FeaturePane::TrackTable:
+            //qDebug() << "LibraryPaneManager::bindLibraryWidget:TrackTable";
+            WLibrary* lib = qobject_cast<WLibrary*>(m_pPaneWidget);
```
We can avoid this cast in a subclass solution

daschuer commented on src/library/librarypanemanager.cpp:
```diff
@@ -0,0 +1,123 @@
+                return;
+            }
+            for (LibraryFeature* f : m_features) {
+                f->bindPaneWidget(lib, pKeyboard);
```
I think we can unify (deduplicate) the feature behavior if we pass here "this", and move some common code to the 
LibraryPaneManager class.

jmigual replied:
With the subclass solution is this necessary? What common code would you move to the LibraryPaneManager class?

daschuer commented on src/library/autodj/dlgautodj.cpp:
```diff
@@ -81,45 +50,40 @@ DlgAutoDJ::DlgAutoDJ(QWidget* parent,
+    pTrackTableView->loadTrackModel(m_pAutoDJTableModel);
+    
+    connect(pTrackTableView, SIGNAL(loadTrack(TrackPointer)),
+            this, SIGNAL(loadTrack(TrackPointer)));
```
These signal to signal connections are looking like a refactoring candidate.

daschuer commented on src/library/autodj/dlgautodj.cpp:
```diff
@@ -191,19 +155,17 @@ void DlgAutoDJ::autoDJStateChanged(AutoDJProcessor::AutoDJState state) {
-    QModelIndexList indices = m_pTrackTableView->selectionModel()->selectedRows();
-
+    LibraryView* pView = m_pLibrary->getActiveView();
+    WTrackTableView* pTrackTable = dynamic_cast<WTrackTableView*>(pView);
```
Is this code correct? We should avoid such casts. 
probably the whole function should be moved to the library back-end or at lest to WTrackTableView

jmigual replied:
No, it is not correct, it was an easy solution because it wasn't working anything and I forgot to do a more consistent code after fixing what was not working :disappointed:

jmigual replied:
The problem is that a feature has now way to know the relation between a pane `ID` and the widget related to this `ID`...

daschuer replied:
Mmm, can't we introduce this relation?

daschuer commented on src/widget/wbaselibrary.cpp:
```diff
@@ -0,0 +1,64 @@
+void WBaseLibrary::switchToView(const QString& name) {
+    QMutexLocker lock(&m_mutex);
+    qDebug() << "WBaseLibrary::switchToView" << name;
+    QWidget* widget = m_viewMap.value(name, nullptr);
```
It looks like we can get rid of m_viewMap and the widget naming if we just pass a QWidget*

jmigual replied:
I think that this was designed like this to allow a more evident separation between the frontend and the backend... All features use the `switchToView(QString)` signal and changing it implies changing many lines of code only to remove the map...

daschuer replied:
Maybe I missed something ... Is there a reason for WBaseLibrary announce every possible view before it  is needed? 

If not, the code will be more straight forward if we pass the widget to the WBaseLibrary when it should be displayed. 

This is will also clarify the lifetime of the pane widgets. Currently they are created by the feature, but destroyed by the Qt widget stack at the very end of Mixxx or if the skin changes. 

It "might" be better if the feature has a strong reference to is views and passes it to the the container widget only on demand. It may survive a skin change. 

However, I do not now if there are style sheet issues with that approach.

jmigual commented:
Yes I plan to distinguish the click case from the drop case

daschuer commented:
The button bar size issue is fixed in LateNight, but not in Deere. The icons are not centered in the German Deere version.  Could you reproduce it?

daschuer commented:
The Library view is still re-randomized when switching between Auto DJ and Library feature using the sidebar buttons.

daschuer commented:
The Drag -> Button hover -> Drop feature works great :-) 

Dropping from one Track table, to an other does not work.

jmigual commented:
Ok, now the re-randomizing bug should not happen :blush:

jmigual commented:
Oh and I've added the new design to Shade skin too

daschuer commented:
Cool B-) Thank you.

Shade: The splitter between cover Art and LibraySidebarExpanded is missing. 

I think the features need to be aware of a hidden track table. 
If the second track table is hidden (or size below a reasonable value), the focus should always stick on the visible pane. 

What should be the default library view? 
IMHO the LibrarySidebar expanded should show the library feature. 
What should be on the second pane? The Auto DJ o

daschuer commented:
It would be nice, if 2.0 skins are still usable. Would it be possible to replace the old treeview from 2.0 skins with the LibrarySidebar + LibrarySidebarExpandned?

daschuer commented:
I just found this bug: 
https://bugs.launchpad.net/mixxx/+bug/1444328

How about two views for AutoDJ, 1 the control buttons, 2 a drop region for crates. 
And a small button or tabs to switch the views.
The splitter looks cluttered.

jmigual commented:
No, it won't be possible to replace the old treeview from 2.0 skins with the LibrarySidebar + LibrarySidebarExpanded because many features (e.g. AutoDJ) have their buttons in the LibrarySidebarExpanded, if we replace this, we will be removing the buttons too. :cry: 

I like your idea for the two AutoDJ views, i totally agree that the splitter is not the best solution and this morning I was thinking how could we solve this. Tabs seem a interesting idea. Do you want too to implement the `Track Sources` description from [this](https://bugs.launchpad.net/mixxx/+bug/1444328/comments/3) comment?

daschuer commented:
> No, it won't be possible to replace the old treeview from 2.0 skins with the LibrarySidebar + LibrarySidebarExpanded because many features (e.g. AutoDJ) have their buttons in the LibrarySidebarExpanded, if we replace this, we will be removing the buttons too. :cry:

Yes, but If we introduce an 2.0 adaptor widget, which contains the LibrarySidebar + LibrarySidebarExpanded and which instantiated by the old LibrarySidebar tag, it might work. 
The new LibrarySidebar can become LibrarySidebarButtons. The same for the old Library which which is search bar + LibraryPane ...
What will prevent it? There is no need for a perfect usability, it would be helpful if it "just works" somehow.  

> I like your idea for the two AutoDJ views, i totally agree that the splitter is not the best solution and this morning I was thinking how could we solve this. Tabs seem a interesting idea. Do you want too to implement the Track Sources description from this comment?

The comment is a good start, but should be tweaked for the new situation. A crate drop container would be nice.

jmigual commented:
> Yes, but If we introduce an 2.0 adaptor widget, which contains the LibrarySidebar + LibrarySidebarExpanded and which instantiated by the old LibrarySidebar tag, it might work. 
The new LibrarySidebar can become LibrarySidebarButtons. The same for the old Library which which is search bar + LibraryPane ...
What will prevent it? There is no need for a perfect usability, it would be helpful if it "just works" somehow.

Ok, I understand it, this can work.

daschuer commented:
Great, an important step for a smooth transition to the new layout. 

None of the search bars in a 2.0 skin is working, is there fake pane id or something missing?

jmigual commented:
Yes there is a fake pane id, and I forgot to pass this id to the library to handle the searchbar from the track table. Now it's solved, I added too the tabs in AutoDJ if you want to try it.

daschuer commented:
Thank you. 

The "drop target" tab should be called more sensible. How about "Track source"

The 2.0 skins are facing segfault when switching to AutoDJ feature: 

Program received signal SIGSEGV, Segmentation fault.
0x000000000094fb55 in LibraryPaneManager::setFocus (this=0x0)
    at src/library/librarypanemanager.cpp:73
73	    m_pPaneWidget->setProperty("showFocus", 1);
(gdb) bt
#0  0x000000000094fb55 in LibraryPaneManager::setFocus (this=0x0)
    at src/library/librarypanemanager.cpp:73
#1  0x000000000094556c in Library::handleFocus (this=this@entry=0x32228a0)
    at src/library/library.cpp:545
#2  0x000000000094618c in Library::slotActivateFeature (
    this=this@entry=0x32228a0, featureName=...) at src/library/library.cpp:422
#3  0x000000000095c90d in Library::qt_static_metacall (_o=0x32228a0, 
    _c=<optimized out>, _id=<optimized out>, _a=0x7fffffffcf20)
    at lin64_build/library/moc_library.cc:118
#4  0x00007ffff51e187a in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x0000000000bbcec0 in WFeatureClickButton::clicked (this=<optimized out>,

jmigual commented:
Ok, thanks for the comment ;-) I'll fix as soon as possible

jmigual commented:
Fixed 😊

daschuer commented:
For our LibraryFeature Plug-In preparation, it looks like we need to just change some minor namings and interfacing things. 

I think we have to change bindPaneWidget(). 
Unfortunately the Fetaure cannot own it's own pane widgets, because of the Qt object / widget tree. 
So we should move to a factory like schema. 
bindPaneWidget() might become a Widget* createPaneWidget() to clarify that the feature passes a widget to the skin widget tree. The named reference should be repaced by a pointer reference.

We should also look at all signal slot connections, maybe we can replace some of them by direct calls.

jmigual commented:
Sorry I don't understant... To me it seems that no feature owns a widget, it only instantiates it in `bindPaneWidget()` and sets the needed parent passed as a parameter. With `createPaneWidget()` the parent will always be null?

daschuer commented:
Yes, that was confusing, sorry.

The general way of operation is not bad. The pane widget stack is already the correct owner. 
My issue, is that the interface and the relations are hard to understand from the functions names and the call hierarchy. It is worth to think about how to improve it. 

The naming could be improved. We should make sure the class and function names are matching our concepts and are self explaining. 

How does the skin library widget relation work now? 
* the skin initializes a LibraryPane stack (WLibrary)
* it gets the LibraryPaneManager from the Library
* it gets the the LibraryFetaures
* the widgets stack is populated by the widgets of the Feature, 

The function is currently called bindPaneWidget(), but that is only half the truth.
A long name would be: 
"CreateThePaneWidgetsOfTheLibraryAddItToTheWLibraryStackAndBindItToTheFeature"
A bit unhandy ;-) 

How about populatePaneWidget()?

I have also not understood where the WTrackTable is bound? Does it make sense to handle it from the same function?  
 
We may get rid of the WBaseLibrary::m_viewMap since it looks like it mirrors a list that is already inside QStackedWidget, which can be used by QStackedWidget::setCurrentWidget(). 

How do you think about it?

jmigual commented:
> How about populatePaneWidget()?

It's ok for me it also describes better what the function does, in the beginning it was hard for me to understand what `bindPaneWidget()` was doing.

> I have also not understood where the WTrackTable is bound? Does it make sense to handle it from the same function?

The `WTrackTable` is bound in the `bindPaneWidget()` function from the library, IMHO it has no sense to be here, it's better to bind the `WTrackTable` in the `bindPaneWidget()` function of the `LibraryPaneManager` or directly in the `LegacySkinParser` (I prefer the first option). 

> We may get rid of the WBaseLibrary::m_viewMap since it looks like it mirrors a list that is already inside QStackedWidget, which can be used by QStackedWidget::setCurrentWidget().

We can't remove the `WBaseLibrary::m_viewMap` because the `QStackedWidget` does not allow to use `QString` as an identifier and all the `switchToView()` signals use this identifier and the `WLibrary` uses it too.

daschuer commented:
> We can't remove the WBaseLibrary::m_viewMap because the QStackedWidget does not allow to use QString as an identifier and all the switchToView() signals use this identifier and the WLibrary uses it too. 

It is Ok to me to keep it, if we have a strong reason. But removing it along with replacing some of these signal slot connections will help reading the code. You are probably deeper in the code then me, so be warned that my refactoring ideas might be rubbish ;-) Her it is: 

The feature populates the pane widget stack as needed. It passed the onership to the skin (widget stack) but holds a (somehow) weak string reference to its pane widget.

This means, he cannot access is own widgets by this string. The feature is not aware of the lifetime of the widget. All communications have to be done by hard to follow signal slot connections or string references. (Which are not tracked by the nice Eclipse "Show call herarchy" feature.) 

Normally I would resolve such a relation by a strong and weak smart pointer pair, but this does not work with the Qt object tree. But we may introduce our own "weak" pointer, watching the QObject::destroyed() signal .

jmigual commented:
Ok now I understand your point of view. When a feature needs to have access to its own widgets currently stores the pointer it passes to the `WLibrary`. With the new panes layout if a feature wants to store this pointer it must also not store one but `n` pointers (one for each pane). This currently can be seen in `AutoDJFeature`, it has a map with `paneID => table pointer`. 
Now you've made me notice that this is not a good behavior and you're right. So... I think that the best way to do this would be a direct `Library <=> Feature` interaction without signals, slots, strings, maps... What do you think about it, each Feature will have a pointer to the Library allowing this direct interaction.

jmigual commented:
A feature can request to the Library their widgets from a certain pane. Then to do this the `bindPaneWidget` should be changed a lot and become a `Widget* createPaneWidget()` the returned widget would have their parent to a `nullptr` and the `LibraryPaneManager` (which called the create pane function) would set the properly parent to the widget (with the `setParent(QObject*)` function from `QObject`). 

This changes many things but I think that is the proper way to do it and to remove some extra code and signals currently used... 
What do you think about it? Is this your first idea with the `createPaneWidget()` function?
Now that I think about it it seems to me that it can work and become more readable code.

daschuer commented:
> What do you think about it?, each Feature will have a pointer to the Library allowing this direct interaction.

Yes, it sounds good. There is "only" a big segmentation fault risk, when the skin has changed. 
Since you have to parent the pane widgets to the pane stack, they will be destroyed when the pane stack is destroyed. That's why we might want to watch the QObject::destroyed() signal, or any other function call that tells the feature that it has just lost one of it's views. 

> Then to do this the bindPaneWidget should be changed a lot and become a Widget* createPaneWidget()

Dose it mean we have only a single pane view per library feature and widget stack? 
A feature with two views like the html view and the track table could return a stack with both views .....
... or we keep the interface in a way that a view can instantiate N views. ....
... I have currently no preference. 

> Is this your first idea with the createPaneWidget() function?

I did not have thought enough about it ;-) It looks probably nice but might be to limiting, I do not know.

jmigual commented:
> Yes, it sounds good. There is "only" a big segmentation fault risk, when the skin has changed. 
Since you have to parent the pane widgets to the pane stack, they will be destroyed when the pane stack is destroyed. That's why we might want to watch the QObject::destroyed() signal, or any other function call that tells the feature that it has just lost one of it's views.

I was thinking in something like requesting the feature widget to the library, something like this

```c++
// m_pLibrary is the pointer to the Library class
m_pLibrary->getPaneWidget(this)->someWidgetFunction()
```

> Dose it mean we have only a single pane view per library feature and widget stack? 
A feature with two views like the html view and the track table could return a stack with both views .....
... or we keep the interface in a way that a view can instantiate N views. ....
... I have currently no preference.

Yes it will be only a single pane per LibraryFeature and widget stack + the track table added by default.
If a widget wants more than one view it can create it's own stack and return it but currently there's no widget using more than one view + track table.

daschuer commented:
> I was thinking in something like requesting the feature widget to the library, something like this

> ```c++
// m_pLibrary is the pointer to the Library class
m_pLibrary->getPaneWidget(this)->someWidgetFunction()
```
Yes, this will work. Maybe like that: 

```c++
// m_pLibrary is the pointer to the Library class
m_pLibrary->getCurrentPaneWidget(this)->someWidgetFunction()
```
to move the n Pane problem from the feature? 

However, I had originally thought of storing the pane pointers in the feature itself.

daschuer commented:
> If a widget wants more than one view it can create it's own stack and return it but currently there's no widget using more than one view + track table.

That sound like a nice improvement.

jmigual commented:
> Yes, this will work. Maybe like that:
```c++
// m_pLibrary is the pointer to the Library class
m_pLibrary->getCurrentPaneWidget(this)->someWidgetFunction()
```

Yes you're right better naming. I was referring to this but I messed it up with a wrong name...

> However, I had originally thought of storing the pane pointers in the feature itself.

If we store the pane pointers then is when we can have bad references caused by pointers deleted on a skin change. I we handle all the pointers in the Library class then it should be safer.

daschuer commented on src/library/analysisfeature.cpp:
```diff
@@ -58,34 +60,74 @@ QIcon AnalysisFeature::getIcon() {
+    	for (auto it = m_analysisTables.begin(); it != m_analysisTables.end(); ++it) {
+    		m_pAnalysisView->setAnalysisTableView(it.value(), it.key());
+    	}
+    	m_analysisTables.clear();
```
This is still hard to understand. 

How about keeping the list of Analysis table m_analysisTables here and pass all calls WAnalysisLibraryTableView <->  DlgAnalysis though the AnalysisFeature. 
It looks like we can get rid of a lot of signal connections this way. 

Did the DlgAnalysis need to know something about the the WAnalysisLibraryTableView views? 
Probably not, if he passes all calls though the AnalysisFeature. 

Currently the load track signal is issued from the WAnalysisLibraryTableView passed though the DlgAnalysis and 
AnalysisFeature and the Library to PlayerManager. This should be replaced by direct calls towards Library.

jmigual replied:
Yes it's a better solution. I'll move all the calls, I supose that this applies to AutoDJ, and MixxxLibrary features too because they use the same method.

daschuer replied:
Yes. Thank you.

daschuer commented on src/widget/wbreadcrumb.h:
```diff
@@ -0,0 +1,23 @@
+    
+  private:
+    
+    static QString& getData(TreeItem* pTree);
```
this function can be moved to the an anonymous namespace in the cpp file.

daschuer commented on src/widget/wbreadcrumb.cpp:
```diff
@@ -0,0 +1,32 @@
+    setText(text);
+}
+
+QString& WBreadCrumb::getData(TreeItem* pTree) {
```
Interesting case, a non const reference to a local variable. 

If you write later like this, it will make sense 
QString& next = getData(pTree->parent());
 
But I am quite sure the compiler has a return value optimization and will optimize the code to get entirely rid of the intermediate object, If you just return the string by value.

daschuer replied:
do we have a more self-explaining name for this function 
getPathString() or something?

jmigual replied:
yes, `getPathString()` is a better name to what it does. I'll fix the reference issue I missed it...

daschuer commented on src/widget/wbreadcrumb.cpp:
```diff
@@ -0,0 +1,32 @@
+QString& WBreadCrumb::getData(TreeItem* pTree) {
+    // Base case
+    if (pTree == nullptr) {
+        return "";
```
return QString()

daschuer commented on src/widget/wbreadcrumb.cpp:
```diff
@@ -0,0 +1,32 @@
+    // Recursive case
+    QString text = pTree->data().toString();
+    QString next = getData(pTree->parent());
+    return (next == "" ? text : next + " > " + text);
```
Here some other "effective" optimisations :-) 

return (next.IsEmpty() ? text : next % QLatin1Literal(" > ") % text);

jmigual replied:
Thanks :wink:

daschuer commented on src/widget/wlibrarystack.cpp:
```diff
@@ -0,0 +1,68 @@
+}
+
+void WLibraryStack::onSearch(const QString& text) {
+    LibraryView* pView = dynamic_cast<LibraryView*>(currentWidget());
```
qobject_cast 

In general we should avoid up-casting, but I think it is OK in this case.

daschuer commented on src/widget/wlibrarystack.cpp:
```diff
@@ -0,0 +1,69 @@
+}
+
+void WLibraryStack::onSearch(const QString& text) {
+    LibraryView* pView = dynamic_cast<LibraryView*>(currentWidget());
```
qobject_cast

In general, we should avoid up-casting but in this case, it should be OK for now.

daschuer commented on src/library/mixxxlibraryfeature.cpp:
```diff
@@ -162,16 +250,57 @@ void MixxxLibraryFeature::refreshLibraryModels() {
+}
+
+void MixxxLibraryFeature::selectAllHidden() {
+    if (m_hiddenPane.contains(m_featureFocus)) {
```
This is a double hash lookup. 
You should use QHash::find instead

daschuer commented on src/library/mixxxlibraryfeature.cpp:
```diff
@@ -162,16 +250,57 @@ void MixxxLibraryFeature::refreshLibraryModels() {
+
+void MixxxLibraryFeature::missingSelectionChanged(const QItemSelection&,
+                                                  const QItemSelection&) {
+    WTrackTableView* pTable = m_missingPane[m_featureFocus];
```
m_featureFocus is -1 [] retuns a default constructed WTrackTableView. 
I think you should us find() as well. The same above.

daschuer commented on src/library/dlganalysis.cpp:
```diff
@@ -167,6 +101,21 @@ int DlgAnalysis::getNumTracks() {
+
+void DlgAnalysis::setSelectedIndexes(const QModelIndexList& selectedIndexes) {
+    qDebug() << "DlgAnalysis::setSelectedIndexes" << selectedIndexes;
+    m_selectedIndexes = selectedIndexes;
```
Is m_selectedIndexes depending on the focus pane?

jmigual replied:
Yes it is, I'll add a comment

daschuer commented on src/library/analysisfeature.cpp:
```diff
@@ -99,6 +145,13 @@ void AnalysisFeature::refreshLibraryModels() {
+void AnalysisFeature::selectAll() {
+    auto it = m_analysisTables.find(m_featureFocus);
+    if (it != m_analysisTables.end()) {
+        (*it)->selectAll();
```
Interesting, I did not know that this actually works. 
Does an explicit it.value()->selectAll(); add any value? 
Probably not, so I am willing to learn. ;-)

jmigual replied:
Using `it.value()->selectAll();` will have the same effect :wink:

daschuer commented on src/library/setlogfeature.cpp:
```diff
@@ -11,10 +11,11 @@
 
-SetlogFeature::SetlogFeature(QObject* parent,
-                             UserSettingsPointer pConfig,
+SetlogFeature::SetlogFeature(UserSettingsPointer pConfig,
```
I think it is the right time to rename this to HistoryFeature. 
(It was my fault not to rename the class when we have changed the visible name :-* )

jmigual replied:
Changed :wink:

daschuer commented:
It is a real pleasure to see  the breadcrumb working. Good work :-) 

The focus handling works for the AutoDJ + Library situation, but not as I would expect in combinations of the other panes. This is probably due to the demand to allow two panes of the same feature side by side. 

How about tweak it a bit like this: 
* Enable AutoDJ in pane 1. 
* Focus  pane 2
* Enable Playlist in pane 2.
* Focus pane 1 (AutoDJ)
* Enable Playlist fetaure
** currently: playlist is enabled in pane 1 as well 
** better: just show the Playlists in SidbareExpanded and switch focus to pane2

If the user wants to enable a second playlist, he can already switch the focus and select a new playlist from the SidbareExpanded.

This approach will break the single pane view. This can solved by never switch the focus to a pane that is not visible. 

What do you think?

jmigual commented:
So with this if the user wants to view the playlists in two panes at once it must do:

1. Focus pane 1
2. Enable playlists (playlists is shown in Pane 1)
3. Select playlist "P1" ("P1" is shown in Pane 1)
4. Focus pane 2
5. Select playlist "P2" ("P2" is shown in Pane 2)

and the user can only select two playlists at once from the Playlist sidebar and not from the Button bar is it?

jmigual commented:
This only applies for Playlists and Crates is it?

daschuer commented:
Yes. 

> This only applies for Playlists and Crates is it?

IMHO when clicking on the sidebar button and that features's pane is already displayed  the focus should always move to this old pane.  That will unify the behavior throughout all features.

jmigual commented:
Ok, so with this change no feature will show in two panes at once  is it?

daschuer commented:
> Ok, so with this change no feature will show in two panes at once is it?

Yes.

jmigual commented:
Ok, if you want to test it, it's done

daschuer commented:
The focus should never move to a hidden pane, collapsed by a slider. 
I think you need to overwrite the QWidget::resizeEvent.

daschuer commented on src/library/librarypanemanager.cpp:
```diff
@@ -0,0 +1,147 @@
+}
+
+void LibraryPaneManager::slotShowBreadCrumb(TreeItem *pTree) {
+    m_pBreadCrumb->showBreadCrumb(pTree);
```
m_pBreadCrumb can be null;

daschuer commented on src/library/librarypanemanager.h:
```diff
@@ -0,0 +1,92 @@
+    WBaseLibrary* m_pPaneWidget;
+    QList<LibraryFeature*> m_features;
+    QHash<LibraryFeature*, int> m_featuresWidget;
+    WLibraryBreadCrumb* m_pBreadCrumb;
```
All the Skin owned widgets pointers need to become QPointers. 
Before using, a null check is required.

daschuer commented:
The Shade skin has no breadcrumbs.

daschuer commented on src/library/libraryfeature.h:
```diff
@@ -4,109 +4,129 @@
+    // Reimplement this to register custom views with the library widget
+    // at the right pane.
+    // TODO(jmigual): This function should be removed and replaced by the
+    // createPaneWidget function
```
It looks like today is the day to do that :-)

daschuer commented on src/library/analysisfeature.cpp:
```diff
@@ -153,23 +213,35 @@ void AnalysisFeature::slotProgressUpdate(int num_left) {
+    }
+    
+    const QModelIndexList &indexes = (*it)->selectionModel()->selectedIndexes();
+    m_pAnalysisView->setSelectedIndexes(indexes);
```
You have to check isNull() every time you use a QPointer

daschuer commented on src/library/analysisfeature.h:
```diff
@@ -23,19 +23,27 @@ class TrackCollection;
 
     QVariant title();
     QIcon getIcon();
+    inline QString getViewName() { 
```
this is a virtual override function which cannot be inlined. Please remove the inline keyword and add the override keyword.

daschuer commented on src/library/analysisfeature.cpp:
```diff
@@ -58,35 +64,79 @@ QIcon AnalysisFeature::getIcon() {
+            this, 
+            SLOT(tableSelectionChanged(const QItemSelection&, const QItemSelection&)));
+    
+    connect(pTable, SIGNAL(loadTrack(TrackPointer)),
```
It would be nice, if we could get rid of this and the following connections. 
That can be done, by passing pLibrary to pTable. Than pTable can directly call the Libray function, that issues a single signal.

jmigual replied:
IMHO this should not be done like this because the direct relation between Library and the Table is not always necessary. In this case it is but if another feature wants to catch the signal before it goes to the Library then this feature won't be able to do it.

daschuer commented on src/library/analysisfeature.h:
```diff
@@ -72,7 +83,9 @@ class AnalysisFeature : public LibraryFeature {
-    DlgAnalysis* m_pAnalysisView;
+    QPointer<DlgAnalysis> m_pAnalysisView;
+    QPointer<AnalysisLibraryTableModel> m_pAnalysisLibraryTableModel;
+    QHash<int, WAnalysisLibraryTableView*> m_analysisTables;
```
This should become a QHash<int, QPointer<WAnalysisLibraryTableView> >
And all -> should be guarded by isNull()

daschuer commented on src/library/libraryfeature.h:
```diff
@@ -4,109 +4,129 @@
+    
+    // Reimplement this to register custom views with the library widget,
+    // at the sidebar expanded pane
+    virtual void bindSidebarWidget(WBaseLibrary *pSidebarWidget,
```
This and seams to be unused as well.

daschuer commented on src/library/analysisfeature.cpp:
```diff
@@ -58,35 +64,79 @@ QIcon AnalysisFeature::getIcon() {
+    
+    m_pAnalysisView->setTableModel(m_pAnalysisLibraryTableModel);
+    
     connect(m_pAnalysisView, SIGNAL(analyzeTracks(QList<TrackId>)),
```
We can get rid of these connections by passing this to m_pAnalysisView

daschuer commented on src/library/libraryfeature.cpp:
```diff
@@ -1,23 +1,82 @@
+    m_featureFocus = focus;
 }
 
 QStringList LibraryFeature::getPlaylistFiles(QFileDialog::FileMode mode) {
```
This function seams to be misplaced. ... but no Idea if we have better place.

jmigual replied:
What do you mean? Do you mean that since `LibraryFeature` is an interface it cannot have a parameter?

daschuer replied:
This is valid code, so not a real issue here. I have just not expected to find such code here since is not an classic "inteface" function which is overridden or something. It is more an independent helper function. We may move it to the Library class, or to its own static class. 
... or just leave th code untouched. 
Since it was already here, it is OK to not change the code.

jmigual replied:
Ok, this was added in PR #945 because some Features inheriting `LibraryFeature` (`PlaylistsFeature` and `CratesFeature`) needed it

daschuer commented on src/library/libraryfeature.cpp:
```diff
@@ -26,14 +85,16 @@ QStringList LibraryFeature::getPlaylistFiles(QFileDialog::FileMode mode) {
             QDesktopServices::storageLocation(QDesktopServices::MusicLocation));
 
     QFileDialog dialogg(NULL,
```
please remove the second "g". I do not know what it was for.

daschuer commented on src/library/librarypanemanager.cpp:
```diff
@@ -0,0 +1,147 @@
+    connect(m_pPaneWidget, SIGNAL(focused()),
+            this, SIGNAL(focused()));
+
+    WLibrary* lib = qobject_cast<WLibrary*>(m_pPaneWidget);
```
"lib" is a bit short how about "wLibrary"

daschuer commented on src/widget/wlibrary.cpp:
```diff
@@ -9,25 +9,21 @@
-        return false;
-    }
+    //qDebug() << "WLibrary::registerView" << name;
     if (dynamic_cast<LibraryView*>(view) == nullptr) {
```
Cant't we avoid that by using WLibrary::registerView(QString name, LibraryView* view) {

jmigual replied:
Since it inherits `WBaseLibrary::registerView(QString name, QWidget* view)` it cannot be done, another option is simply adding this since this is only called in a single line of code inside `LibraryPaneManager`

jmigual replied:
Also it needs to be a `QWidget*` because it will be added to the `QStackedWidget`

daschuer commented on src/widget/wlibrary.cpp:
```diff
@@ -9,25 +9,21 @@
 }
 
 bool WLibrary::registerView(QString name, QWidget* view) {
     QMutexLocker lock(&m_mutex);
```
"name" is now unused.

daschuer commented on src/library/libraryfeature.h:
```diff
@@ -4,109 +4,129 @@
+    
+    // Must be a unique name for each feature, it must be a unique name for each
+    // different feature
+    virtual QString getViewName() = 0;
```
getFeatureName()

daschuer commented on src/widget/wfeatureclickbutton.cpp:
```diff
@@ -0,0 +1,71 @@
+
+    setIcon(m_feature->getIcon());
+    setText(m_feature->title().toString());
+    m_data = m_feature->getViewName();
```
I think we can store the feature pointer here instead.

jmigual replied:
Yes, it will be more usefull

daschuer commented on src/widget/wfeatureclickbutton.cpp:
```diff
@@ -0,0 +1,71 @@
+        QToolButton::timerEvent(event);
+        return;
+    }
+    emit(hoverShow(m_data));
```
Than can become 
emit(hoverShow(m_feature->getFeatureName());

jmigual commented:
Oh yes I forgot the breadcrumbs in shade skin!

jmigual commented:
Ok, now Shade skin has the breadcrumb and the focus never moves to a hidden pane, and the pointers issues are fixed.

daschuer commented on src/widget/wbaselibrary.cpp:
```diff
@@ -0,0 +1,79 @@
+    return true;
+}
+
+QString WBaseLibrary::getCurrentViewName() {
```
This the name of the feature that is associated with the view. 
So it is something like getCurrentViewsFeatureName() 

However, I like to get rid of all this Naming and string stuff. It introduces just an other layer of complexity, without a real benefit.

jmigual replied:
Yeah this is very confusing... But this problem appears because the WTrackTable is not a feature but it's included in the WBaseLibrary, and this causes all the problems... Maybe separating it into another stack would solve the problem so it will be like this:

- QStackedWidget
    - WTrackTable
    - WBaseLibrary
        - Library widget 1
        - Library widget 2

What do you think about it? This should solve the problem.

daschuer replied:
Yes, if I understand the issue correct, every feature should handle it's on views. So need a separate WTrackTable for every feature. This will help also to save the scroll position when changing the features.

- QStackedWidget
    - WBaseLibrary
        - Library widget 1
            - WTrackTable 1
        - Library widget 2
            - WTrackTable 2

daschuer commented on src/library/itunes/itunesfeature.cpp:
```diff
@@ -112,6 +117,10 @@ QIcon ITunesFeature::getIcon() {
+}
+
 void ITunesFeature::activate() {
     activate(false);
```
Does this function already works for the new schema? 

m_pLibrary->slotShowTrackModel(m_pITunesTrackModel);
Didn't we discuss, to have a spearate WTrackTable for every feature?

jmigual replied:
Yes this function works for the new schema but it gives to me many headaches so adding a separate WTrackTable for every feature will solve all of this headaches :blush:

daschuer commented on src/library/autodj/autodjfeature.cpp:
```diff
@@ -94,28 +98,59 @@ QIcon AutoDJFeature::getIcon() {
+    QScrollArea* pScroll = new QScrollArea(pContainer);
+    pScroll->setWidget(m_pAutoDJView);
+    pScroll->setWidgetResizable(true);
+    pContainer->addTab(pScroll, tr("controls"));
```
Please switch to "controls" Tab by default. Use capitalized "Controls".

daschuer commented:
> Avoid moving focus to a hidden pane

works great now :+1: . Thank you

daschuer commented on src/library/analysisfeature.cpp:
```diff
@@ -58,58 +63,83 @@ QIcon AnalysisFeature::getIcon() {
+    m_pLibrary->slotShowBreadCrumb(m_childModel.getItem(QModelIndex()));
+    
+    if (!m_pAnalysisView.isNull()) {
         emit(restoreSearch(m_pAnalysisView->currentSearch()));
```
This can become a direct call m_pLibrary->slotRestoreSearch(m_pAnalysisView->currentSearch()))

daschuer commented on src/library/librarypanemanager.cpp:
```diff
@@ -0,0 +1,179 @@
+    m_pPaneWidget->setCurrentIndex(widgetId);
+}
+
+void LibraryPaneManager::slotRestoreSearch(const QString& text) {
```
It looks like this one is unused and can be removed.

jmigual replied:
This is still used but it's not an slot anymore. It's called form the Library class when the slot restore search is called.

jmigual replied:
I think that I'll try to remove the called signal and avoid unnecessary code too

daschuer commented:
Thank you for removing all unnecessary signal slot connections. 
Be sure to remove the "slot" prefix from the function names (using Eclipse refactoring). 
Remove the ex-slot functions from the slot region in the headers and delete the unused signals.

daschuer commented on src/library/libraryfeature.h:
```diff
@@ -4,113 +4,132 @@
+    
+    // Must be a unique name for each feature, it must be a unique name for each
+    // different feature
+    virtual QString getFeatureName() = 0;
```
It looks like this and its overrides are finally unused :-)

jmigual replied:
Yes! And I forgot to remove it 😓

On Sun, Jul 3, 2016, 12:19 Daniel Schürmann <notifications@github.com>
wrote:

> In src/library/libraryfeature.h
> <https://github.com/daschuer/mixxx/pull/10#discussion_r69388910>:
>
> >      virtual ~LibraryFeature();
> >
> >      virtual QVariant title() = 0;
> >      virtual QIcon getIcon() = 0;
> > +
> > +    // Must be a unique name for each feature, it must be a unique name for each
> > +    // different feature
> > +    virtual QString getFeatureName() = 0;
>
> It looks like this and its overrides are finally unused :-)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/daschuer/mixxx/pull/10/files/fd67608147f2b65021b6b0e7ce49ddf7a3f9ae47#r69388910>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGMV00HCev3jAg5TxQrZUrFZuY7Qn-hNks5qR4ycgaJpZM4Ir3iU>
> .
>

daschuer commented:
There is an issue with your own master branch. 
https://github.com/daschuer/mixxx/pull/10/commits/28db1fd87d8886782ba60ea8226397fac91dacc8
Should not exist. 

It looks like it is you intention that your master branch mirrors the mixxxdj/mixxx master branch,  but it is not.   

The issue starts with e5cb15e0ba178dcf5bd4b2e7d69b8d0da62d2da8
This is not a real issue, since the additional commits seams not to contain code, but we should avoid future null merge commits. 

This can be done by resetting your master branch: 

git checkout master 
git reset upstream/master --hard 

You should also remove the latest null merge commit by redo the master merge: 

git checkout feature/libraryManager
git reset b0fd7c7206426a104b840a631c910f0e41784688
git add src/test/durationutiltest.cpp src/util/duration.cpp
git stash 
git merge upstream/master -s ours
git stash pop 
git commit -a --amend
git push -f

daschuer commented on src/library/baseplaylistfeature.h:
```diff
@@ -75,14 +76,17 @@ class BasePlaylistFeature : public LibraryFeature {
     TrackDAO &m_trackDao;
-    PlaylistTableModel* m_pPlaylistTableModel;
+    QPointer<PlaylistTableModel> m_pPlaylistTableModel;
+    QHash<int, QPointer<PlaylistTableModel> > m_playlistTableModel;
```
It is hard to distinguish m_pPlaylistTableModel from m_playlistTableModel. 
How about m_playlistTableModelHash? 

Can m_playlistTableModelHash become private? 

Since both are protected members the lifetime, should be clear. 
Or are they used outside the creating class after the feature is gone?

jmigual commented:
Yes m_playlistTableModelHash can become private. It's access is done only in the getPlaylistTableModel function, I'll change it
